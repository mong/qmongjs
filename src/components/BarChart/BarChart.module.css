.wrapper {
  padding: 0 50px;
  margin-bottom: 2rem;
}

.barChart {
  overflow: visible;
  width: 100%;
  display: block;
}
