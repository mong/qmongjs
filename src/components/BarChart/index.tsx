import {
  axisBottom,
  axisLeft,
  format,
  scaleBand,
  scaleLinear,
  select,
} from "d3";
import React, { useEffect, useRef } from "react";
import useResizeObserver from "../utils";
import styles from "./BarChart.module.css";

export interface DataPoint {
  label: string;
  value: number;
}

export interface Levels {
  mid: number;
  low: number;
}

export interface Props {
  displayLevels: boolean;
  data: DataPoint[];
  levels: Levels;
}

function levelColor(level: string) {
  switch (level) {
    case "high":
      return "#3baa34";
    case "mid":
      return "#fd9c00";
    case "low":
      return "#e30713";
    default:
      throw new Error(`${level} is not a valid level`);
  }
}

function BarChart({ data, displayLevels, levels: { low, mid } }: Props) {
  const svgRef = useRef<SVGSVGElement>(null);
  const wrapperRef = useRef<HTMLDivElement>(null);
  const entry = useResizeObserver(wrapperRef);

  const height = data.length * 50;
  const width = entry?.contentRect.width ?? 0;

  useEffect(() => {
    if (!svgRef.current) {
      return;
    }

    const svg = select(svgRef.current);

    // Scales
    const yScale = scaleBand()
      .domain(data.map((d) => d.label))
      .range([0, height])
      .padding(0.5);

    const xScale = scaleLinear().domain([0, 1]).range([0, width]);

    // Y-Axis
    const yAxis = axisLeft(yScale);
    const yAxisElement = svg.select<SVGGElement>(".y-axis");
    yAxisElement.call(yAxis);
    yAxisElement.select(".domain").remove();
    yAxisElement.selectAll(".tick line").remove();

    // X-Axis
    const xAxis = axisBottom(xScale)
      .tickSize(-height)
      .ticks(6)
      .tickFormat(format(",.0%"));
    const xAxisElement = svg.select<SVGGElement>(".x-axis");
    xAxisElement.style("transform", `translateY(${height}px)`).call(xAxis);
    xAxisElement.select(".domain").remove();
    xAxisElement.selectAll(".tick line").attr("stroke", "#D2D3D4");
    xAxisElement
      .selectAll(".tick text")
      .attr("fill", "#828586")
      .attr("font-size", "18.57px");

    // Levels
    const levels = makeLevels(mid, low);

    svg
      .selectAll(".level")
      .data(Object.entries(levels))
      .join("rect")
      .attr("class", "level")
      .attr("data-testid", ([level]) => `level-${level}`)
      .attr("y", 0)
      .attr("x", ([, range]) => xScale(range.start))
      .attr("width", ([, range]) => xScale(range.end - range.start))
      .attr("height", height)
      .attr("fill", ([level]) => levelColor(level))
      .attr("opacity", "0.2")
      .attr("visibility", displayLevels ? "visible" : "hidden");

    // Bars
    svg
      .selectAll(".bar")
      .data(data)
      .join("rect")
      .attr("class", "bar")
      .attr("data-testid", (d) => `bar-${d.label}`)
      .attr("x", 0)
      .attr("y", (d) => yScale(d.label) ?? 0)
      .attr("height", yScale.bandwidth())
      .attr("fill", "#00263d")
      .transition()
      .attr("width", (d) => xScale(d.value));
  }, [data, width, height, displayLevels, low, mid]);

  return (
    <div ref={wrapperRef} className={styles.wrapper}>
      <svg ref={svgRef} className={styles.barChart} height={height}>
        <g className="x-axis" />
        <g className="y-axis" />
      </svg>
    </div>
  );
}

export default BarChart;

function makeLevels(mid: number, low: number) {
  if (mid < low) {
    return {
      high: { start: 0, end: mid },
      mid: { start: mid, end: low },
      low: { start: low, end: 1 },
    };
  }

  return {
    high: { start: mid, end: 1 },
    mid: { start: low, end: mid },
    low: { start: 0, end: low },
  };
}
