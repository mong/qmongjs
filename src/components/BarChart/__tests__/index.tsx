import { render, waitFor, screen } from "@testing-library/react";
import React from "react";
import { build, fake, bool } from "@jackfranklin/test-data-bot";
import BarChart, { Props, DataPoint } from "..";
import useResizeObserver from "../../utils";
import { Level } from "../../TF_FIGURE/Chart";

jest.mock("../../utils");

/**
 * Move all d3 transitions a specified amount of milliseconds forward in time.
 * @param milliseconds Milliseconds to move transitions forward in time
 */
function clockTick(milliseconds: number) {
  const currentNow = performance.now();

  const now = performance.now;
  performance.now = () => currentNow + milliseconds;
  // The new animation frame means d3's timers will check performance.now() again
  return new Promise((resolve) =>
    requestAnimationFrame((time) => {
      performance.now = now;
      resolve(time);
    })
  );
}

test("Bar widths are correct", async () => {
  const WIDTH = 500;
  (useResizeObserver as jest.Mock).mockReturnValue({
    contentRect: {
      width: WIDTH,
    },
  });

  const props = propsBuilder();

  render(
    <BarChart
      {...props}
      zoom={false}
      margin={{ top: 0, right: 0, bottom: 0, left: 0 }}
    />
  );

  await clockTick(1500);

  for (const dataPoint of props.data) {
    const bar = screen.getByTestId(`bar-${dataPoint.label}`);

    const width = bar.getAttribute("width") ?? "";
    expect(parseFloat(width)).toBeCloseTo(dataPoint.value * WIDTH);
  }
});

test("Level widths are correct", async () => {
  const WIDTH = 500;
  (useResizeObserver as jest.Mock).mockReturnValue({
    contentRect: {
      width: WIDTH,
    },
  });

  const props = propsBuilder();

  render(
    <BarChart
      {...props}
      zoom={false}
      margin={{ top: 0, right: 0, bottom: 0, left: 0 }}
    />
  );

  await clockTick(1500);

  for (const l of props.levels) {
    const level = screen.getByTestId(`level-${l.level}`);

    const levelX = level.getAttribute("x") ?? "";
    expect(parseFloat(levelX)).toBeCloseTo(l.end * WIDTH);

    const levelWidth = level.getAttribute("width") ?? "";
    expect(parseFloat(levelWidth)).toBeCloseTo(l.start * WIDTH - l.end * WIDTH);
  }
});

test("Render without levels @250px", async () => {
  const WIDTH = 250;
  (useResizeObserver as jest.Mock).mockReturnValue({
    contentRect: {
      width: WIDTH,
    },
  });

  const { container } = render(
    <BarChart
      displayLevels={false}
      data={[
        { label: "a", value: 1 },
        { label: "b", value: 0.15 },
        { label: "c", value: 0.3 },
        { label: "d", value: 0.1 },
      ]}
      levels={[
        { level: "high", start: 1, end: 0.09 },
        { level: "mid", start: 0.9, end: 0.5 },
        { level: "low", start: 0.5, end: 0 },
      ]}
      zoom={false}
      margin={{ top: 0, right: 0, bottom: 0, left: 0 }}
    />
  );

  await clockTick(1500);

  expect(container).toMatchSnapshot();
});

test("Render with levels @250px", async () => {
  const WIDTH = 250;
  (useResizeObserver as jest.Mock).mockReturnValue({
    contentRect: {
      width: WIDTH,
    },
  });

  const { container } = render(
    <BarChart
      displayLevels={true}
      data={[
        { label: "a", value: 1 },
        { label: "b", value: 0.15 },
        { label: "c", value: 0.3 },
        { label: "d", value: 0.1 },
      ]}
      levels={[
        { level: "high", start: 1, end: 0.09 },
        { level: "mid", start: 0.9, end: 0.5 },
        { level: "low", start: 0.5, end: 0 },
      ]}
      zoom={false}
      margin={{ top: 0, right: 0, bottom: 0, left: 0 }}
    />
  );

  await clockTick(1500);

  expect(container).toMatchSnapshot();
});

test("Render without levels @500px", async () => {
  const WIDTH = 500;
  (useResizeObserver as jest.Mock).mockReturnValue({
    contentRect: {
      width: WIDTH,
    },
  });

  const { container } = render(
    <BarChart
      displayLevels={false}
      data={[
        { label: "a", value: 1 },
        { label: "b", value: 0.15 },
        { label: "c", value: 0.3 },
        { label: "d", value: 0.1 },
      ]}
      levels={[
        { level: "high", start: 1, end: 0.09 },
        { level: "mid", start: 0.9, end: 0.5 },
        { level: "low", start: 0.5, end: 0 },
      ]}
      zoom={false}
      margin={{ top: 0, right: 0, bottom: 0, left: 0 }}
    />
  );

  await clockTick(1500);

  expect(container).toMatchSnapshot();
});

test("Render with levels @500px", async () => {
  const WIDTH = 500;
  (useResizeObserver as jest.Mock).mockReturnValue({
    contentRect: {
      width: WIDTH,
    },
  });

  const { container } = render(
    <BarChart
      displayLevels={true}
      data={[
        { label: "a", value: 1 },
        { label: "b", value: 0.15 },
        { label: "c", value: 0.3 },
        { label: "d", value: 0.1 },
      ]}
      levels={[
        { level: "high", start: 1, end: 0.09 },
        { level: "mid", start: 0.9, end: 0.5 },
        { level: "low", start: 0.5, end: 0 },
      ]}
      zoom={false}
      margin={{ top: 0, right: 0, bottom: 0, left: 0 }}
    />
  );

  await clockTick(1500);

  expect(container).toMatchSnapshot();
});

test("Render with levels reversed @500px", async () => {
  const WIDTH = 500;
  (useResizeObserver as jest.Mock).mockReturnValue({
    contentRect: {
      width: WIDTH,
    },
  });

  const { container } = render(
    <BarChart
      displayLevels={true}
      data={[
        { label: "a", value: 1 },
        { label: "b", value: 0.15 },
        { label: "c", value: 0.3 },
        { label: "d", value: 0.1 },
      ]}
      levels={[
        { level: "high", start: 0.5, end: 0 },
        { level: "mid", start: 0.9, end: 0.5 },
        { level: "low", start: 1, end: 0.9 },
      ]}
      zoom={false}
      margin={{ top: 0, right: 0, bottom: 0, left: 0 }}
    />
  );

  await clockTick(1500);

  expect(container).toMatchSnapshot();
});

test("Render zoomed with levels @500px", async () => {
  const WIDTH = 500;
  (useResizeObserver as jest.Mock).mockReturnValue({
    contentRect: {
      width: WIDTH,
    },
  });

  const { container } = render(
    <BarChart
      displayLevels={true}
      data={[
        { label: "a", value: 1 },
        { label: "b", value: 0.15 },
        { label: "c", value: 0.3 },
        { label: "d", value: 0.1 },
      ]}
      levels={[
        { level: "high", start: 1, end: 0.09 },
        { level: "mid", start: 0.9, end: 0.5 },
        { level: "low", start: 0.5, end: 0 },
      ]}
      zoom
      margin={{ top: 0, right: 0, bottom: 0, left: 0 }}
    />
  );

  await clockTick(1500);

  expect(container).toMatchSnapshot();
});

// Builders
const dataPointBuilder = build<DataPoint>("DataPoint", {
  fields: {
    label: fake((f) => f.random.uuid()),
    value: fake((f) => f.random.number(100) / 100),
  },
});

const propsBuilder = build<Props>("Props", {
  fields: {
    displayLevels: bool(),
    data: fake((f) =>
      Array.from(
        { length: f.random.number({ min: 1, max: 10 }) },
        dataPointBuilder
      )
    ),
    levels: fake((f): Level[] => {
      const low = f.random.number({ min: 0, max: 100 }) / 100;
      const mid = f.random.number({ min: low * 100, max: 100 }) / 100;

      return [
        { level: "high", start: mid, end: 1 },
        { level: "mid", start: low, end: mid },
        { level: "low", start: 0, end: low },
      ];
    }),
    zoom: bool(),
  },
});
