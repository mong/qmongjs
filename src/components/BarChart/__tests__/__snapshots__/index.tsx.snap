// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render with levels @250px 1`] = `
<div>
  <div
    class="wrapper"
  >
    <svg
      class="barChart"
      height="200"
    >
      <g
        class="x-axis"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        style="transform: translateY(200px);"
        text-anchor="middle"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(0.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            0%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(50.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            20%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(100.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            40%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(150.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            60%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(200.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            80%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(250.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            100%
          </text>
        </g>
      </g>
      <g
        class="y-axis"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="end"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(0,33.33333333333334)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            a
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,77.77777777777779)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            b
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,122.22222222222223)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            c
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,166.66666666666666)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            d
          </text>
        </g>
      </g>
      <rect
        class="level"
        data-testid="level-high"
        fill="#3baa34"
        height="200"
        opacity="0.2"
        visibility="visible"
        width="24.999999999999993"
        x="225"
        y="0"
      />
      <rect
        class="level"
        data-testid="level-mid"
        fill="#fd9c00"
        height="200"
        opacity="0.2"
        visibility="visible"
        width="100"
        x="125"
        y="0"
      />
      <rect
        class="level"
        data-testid="level-low"
        fill="#e30713"
        height="200"
        opacity="0.2"
        visibility="visible"
        width="125"
        x="0"
        y="0"
      />
      <rect
        class="bar"
        data-testid="bar-a"
        fill="#00263d"
        height="22.22222222222222"
        width="250"
        x="0"
        y="22.22222222222223"
      />
      <rect
        class="bar"
        data-testid="bar-b"
        fill="#00263d"
        height="22.22222222222222"
        width="37.5"
        x="0"
        y="66.66666666666667"
      />
      <rect
        class="bar"
        data-testid="bar-c"
        fill="#00263d"
        height="22.22222222222222"
        width="75"
        x="0"
        y="111.11111111111111"
      />
      <rect
        class="bar"
        data-testid="bar-d"
        fill="#00263d"
        height="22.22222222222222"
        width="25"
        x="0"
        y="155.55555555555554"
      />
    </svg>
  </div>
</div>
`;

exports[`Render with levels @500px 1`] = `
<div>
  <div
    class="wrapper"
  >
    <svg
      class="barChart"
      height="200"
    >
      <g
        class="x-axis"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        style="transform: translateY(200px);"
        text-anchor="middle"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(0.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            0%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(100.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            20%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(200.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            40%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(300.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            60%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(400.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            80%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(500.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            100%
          </text>
        </g>
      </g>
      <g
        class="y-axis"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="end"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(0,33.33333333333334)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            a
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,77.77777777777779)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            b
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,122.22222222222223)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            c
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,166.66666666666666)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            d
          </text>
        </g>
      </g>
      <rect
        class="level"
        data-testid="level-high"
        fill="#3baa34"
        height="200"
        opacity="0.2"
        visibility="visible"
        width="49.999999999999986"
        x="450"
        y="0"
      />
      <rect
        class="level"
        data-testid="level-mid"
        fill="#fd9c00"
        height="200"
        opacity="0.2"
        visibility="visible"
        width="200"
        x="250"
        y="0"
      />
      <rect
        class="level"
        data-testid="level-low"
        fill="#e30713"
        height="200"
        opacity="0.2"
        visibility="visible"
        width="250"
        x="0"
        y="0"
      />
      <rect
        class="bar"
        data-testid="bar-a"
        fill="#00263d"
        height="22.22222222222222"
        width="500"
        x="0"
        y="22.22222222222223"
      />
      <rect
        class="bar"
        data-testid="bar-b"
        fill="#00263d"
        height="22.22222222222222"
        width="75"
        x="0"
        y="66.66666666666667"
      />
      <rect
        class="bar"
        data-testid="bar-c"
        fill="#00263d"
        height="22.22222222222222"
        width="150"
        x="0"
        y="111.11111111111111"
      />
      <rect
        class="bar"
        data-testid="bar-d"
        fill="#00263d"
        height="22.22222222222222"
        width="50"
        x="0"
        y="155.55555555555554"
      />
    </svg>
  </div>
</div>
`;

exports[`Render with levels reversed 1`] = `
<div>
  <div
    class="wrapper"
  >
    <svg
      class="barChart"
      height="200"
    >
      <g
        class="x-axis"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        style="transform: translateY(200px);"
        text-anchor="middle"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(0.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            0%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(100.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            20%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(200.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            40%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(300.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            60%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(400.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            80%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(500.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            100%
          </text>
        </g>
      </g>
      <g
        class="y-axis"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="end"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(0,33.33333333333334)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            a
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,77.77777777777779)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            b
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,122.22222222222223)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            c
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,166.66666666666666)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            d
          </text>
        </g>
      </g>
      <rect
        class="level"
        data-testid="level-high"
        fill="#3baa34"
        height="200"
        opacity="0.2"
        visibility="visible"
        width="250"
        x="0"
        y="0"
      />
      <rect
        class="level"
        data-testid="level-mid"
        fill="#fd9c00"
        height="200"
        opacity="0.2"
        visibility="visible"
        width="200"
        x="250"
        y="0"
      />
      <rect
        class="level"
        data-testid="level-low"
        fill="#e30713"
        height="200"
        opacity="0.2"
        visibility="visible"
        width="49.999999999999986"
        x="450"
        y="0"
      />
      <rect
        class="bar"
        data-testid="bar-a"
        fill="#00263d"
        height="22.22222222222222"
        width="500"
        x="0"
        y="22.22222222222223"
      />
      <rect
        class="bar"
        data-testid="bar-b"
        fill="#00263d"
        height="22.22222222222222"
        width="75"
        x="0"
        y="66.66666666666667"
      />
      <rect
        class="bar"
        data-testid="bar-c"
        fill="#00263d"
        height="22.22222222222222"
        width="150"
        x="0"
        y="111.11111111111111"
      />
      <rect
        class="bar"
        data-testid="bar-d"
        fill="#00263d"
        height="22.22222222222222"
        width="50"
        x="0"
        y="155.55555555555554"
      />
    </svg>
  </div>
</div>
`;

exports[`Render without levels @250px 1`] = `
<div>
  <div
    class="wrapper"
  >
    <svg
      class="barChart"
      height="200"
    >
      <g
        class="x-axis"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        style="transform: translateY(200px);"
        text-anchor="middle"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(0.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            0%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(50.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            20%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(100.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            40%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(150.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            60%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(200.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            80%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(250.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            100%
          </text>
        </g>
      </g>
      <g
        class="y-axis"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="end"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(0,33.33333333333334)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            a
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,77.77777777777779)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            b
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,122.22222222222223)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            c
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,166.66666666666666)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            d
          </text>
        </g>
      </g>
      <rect
        class="level"
        data-testid="level-high"
        fill="#3baa34"
        height="200"
        opacity="0.2"
        visibility="hidden"
        width="24.999999999999993"
        x="225"
        y="0"
      />
      <rect
        class="level"
        data-testid="level-mid"
        fill="#fd9c00"
        height="200"
        opacity="0.2"
        visibility="hidden"
        width="100"
        x="125"
        y="0"
      />
      <rect
        class="level"
        data-testid="level-low"
        fill="#e30713"
        height="200"
        opacity="0.2"
        visibility="hidden"
        width="125"
        x="0"
        y="0"
      />
      <rect
        class="bar"
        data-testid="bar-a"
        fill="#00263d"
        height="22.22222222222222"
        width="250"
        x="0"
        y="22.22222222222223"
      />
      <rect
        class="bar"
        data-testid="bar-b"
        fill="#00263d"
        height="22.22222222222222"
        width="37.5"
        x="0"
        y="66.66666666666667"
      />
      <rect
        class="bar"
        data-testid="bar-c"
        fill="#00263d"
        height="22.22222222222222"
        width="75"
        x="0"
        y="111.11111111111111"
      />
      <rect
        class="bar"
        data-testid="bar-d"
        fill="#00263d"
        height="22.22222222222222"
        width="25"
        x="0"
        y="155.55555555555554"
      />
    </svg>
  </div>
</div>
`;

exports[`Render without levels @500px 1`] = `
<div>
  <div
    class="wrapper"
  >
    <svg
      class="barChart"
      height="200"
    >
      <g
        class="x-axis"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        style="transform: translateY(200px);"
        text-anchor="middle"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(0.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            0%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(100.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            20%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(200.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            40%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(300.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            60%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(400.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            80%
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(500.5,0)"
        >
          <line
            stroke="#D2D3D4"
            y2="-200"
          />
          <text
            dy="0.71em"
            fill="#828586"
            font-size="18.57px"
            y="3"
          >
            100%
          </text>
        </g>
      </g>
      <g
        class="y-axis"
        fill="none"
        font-family="sans-serif"
        font-size="10"
        text-anchor="end"
      >
        <g
          class="tick"
          opacity="1"
          transform="translate(0,33.33333333333334)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            a
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,77.77777777777779)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            b
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,122.22222222222223)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            c
          </text>
        </g>
        <g
          class="tick"
          opacity="1"
          transform="translate(0,166.66666666666666)"
        >
          <text
            dy="0.32em"
            fill="currentColor"
            x="-9"
          >
            d
          </text>
        </g>
      </g>
      <rect
        class="level"
        data-testid="level-high"
        fill="#3baa34"
        height="200"
        opacity="0.2"
        visibility="hidden"
        width="49.999999999999986"
        x="450"
        y="0"
      />
      <rect
        class="level"
        data-testid="level-mid"
        fill="#fd9c00"
        height="200"
        opacity="0.2"
        visibility="hidden"
        width="200"
        x="250"
        y="0"
      />
      <rect
        class="level"
        data-testid="level-low"
        fill="#e30713"
        height="200"
        opacity="0.2"
        visibility="hidden"
        width="250"
        x="0"
        y="0"
      />
      <rect
        class="bar"
        data-testid="bar-a"
        fill="#00263d"
        height="22.22222222222222"
        width="500"
        x="0"
        y="22.22222222222223"
      />
      <rect
        class="bar"
        data-testid="bar-b"
        fill="#00263d"
        height="22.22222222222222"
        width="75"
        x="0"
        y="66.66666666666667"
      />
      <rect
        class="bar"
        data-testid="bar-c"
        fill="#00263d"
        height="22.22222222222222"
        width="150"
        x="0"
        y="111.11111111111111"
      />
      <rect
        class="bar"
        data-testid="bar-d"
        fill="#00263d"
        height="22.22222222222222"
        width="50"
        x="0"
        y="155.55555555555554"
      />
    </svg>
  </div>
</div>
`;
